model/jpa/src/main/java/org/keycloak/migration/jpa/LazyCreateIndexCustomChange.java

```diff
@@ -0,0 +1,163 @@
+import liquibase.change.AddColumnConfig;
+import liquibase.change.custom.CustomSqlChange;
+import liquibase.database.Database;
+import liquibase.database.core.MSSQLDatabase;
+import liquibase.database.core.MySQLDatabase;
+import liquibase.database.core.OracleDatabase;
+import liquibase.database.core.PostgresDatabase;
+import liquibase.exception.SetupException;
+import liquibase.exception.ValidationErrors;
+import liquibase.resource.ResourceAccessor;
+import liquibase.statement.SqlStatement;
+import liquibase.statement.core.CreateIndexStatement;
+import liquibase.statement.core.RawSqlStatement;
+import liquibase.structure.core.Column;
+import liquibase.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.StringJoiner;
+
+public class LazyCreateIndexCustomChange implements CustomSqlChange {
+
+    private String catalogName;
+    private String schemaName;
+    private String tableName;
+    private String indexName;
+    private List<AddColumnConfig> columns = Collections.emptyList();
+
+    public String getIndexName() {
+        return indexName;
+    }
+
+    public void setIndexName(String indexName) {
+        this.indexName = indexName;
+    }
+
+    public String getSchemaName() {
+        return schemaName;
+    }
+
+    public void setSchemaName(String schemaName) {
+        this.schemaName = schemaName;
+    }
+
+    public String getTableName() {
+        return tableName;
+    }
+
+    public void setTableName(String tableName) {
+        this.tableName = tableName;
+    }
+
+    @Override
+    public SqlStatement[] generateStatements(Database database) {
+        final String escapedTableName = database.escapeTableName(getCatalogName(), getSchemaName(), getTableName());
+        final String escapedIndexName = database.escapeIndexName(getCatalogName(), getSchemaName(), getIndexName());
+        final String escapedColumnNames = joinEscapedColumnNames(database);
+
+        final String rawQuery;
+        if (database instanceof PostgresDatabase) {
+            rawQuery = "CREATE INDEX " + escapedIndexName +
+                    " ON " + escapedTableName +
+                    " (" + escapedColumnNames + ") CONCURRENTLY";
+        } else if (database instanceof OracleDatabase) {
+            rawQuery = "CREATE INDEX " + escapedIndexName +
+                    " ON " + escapedTableName +
+                    " (" + escapedColumnNames + ") ONLINE";
+        } else if (database instanceof MySQLDatabase) {
+            rawQuery = "ALTER TABLE " + escapedTableName +
+                    " ADD INDEX " + escapedIndexName +
+                    " (" + escapedColumnNames + ")" +
+                    ", LOCK=NONE;";
+        } else if (database instanceof MSSQLDatabase) {
+            rawQuery = "CREATE NONCLUSTERED INDEX " + escapedIndexName +
+                    " ON " + escapedTableName + " (" + escapedColumnNames + ")" +
+                    " REBUILD WITH (ONLINE = ON)";
```

Comment:
Does this work for all MSSQL version? according to their documenation, ONLINE=ON is only supported for Enterprise MSSQL -- which some customers might not have.

Reply:
Yes, I agree, only Enterprise MSSQL supports the online flag. I think the Keycloak core team  should reconsider the conclusion regarding these indexes. There are several issues related to missing indexes and several of them causes real problems. At least a constructive reply would be helpful. Right no it seems like the issues are just ignored.

Reply:
I agree that the missing indices are painful. This issue is not being ignored, however there are deployments that could have been disrupted by forced creation of the indices.

Clearly there is no simple automated way to create the indices due to differences among the databases. Perhaps we can do the following liquibase extension similar to what you have implemented:
* An index can just be created in the foreground if the affected table rows count is smaller than a given configurable number with a reasonable default value (let say 500.000 rows)
* Otherwise a warning with the SQL message is logged that contains a SQL command recommended to be applied manually by the DB admin.

Would that work for you, @thielef, @knutz3n, @gtudan ?

Reply:
+1. Sounds like a good compromise. On top of that the official documentation (+release notes) should get a chapter about this fact with arguments why Keycloak cannot apply this automatically to the whole bunch of existing setups.

Reply:
Yes, I think that is a fair compromise. I have a couple of questions, though.
1. Where should the configurable number be be configured? Standalone.xml?
2. Will Liquibase have knowledge of the index after it is applied manually? Does it need to?
3. Would it be an option to require the DB admin to add the index before upgrading? So Liquibase can just ensure that the index exists and if not, and there are less than x rows, it will apply the index?

Also, regarding Postgres which has support to add indexes in the background, it would be nice to get it to work with adding indexes concurrently. However, Postgres requires this to be done outside an transaction. When I tested it in Keycloak I got a deadlock even when the index was complete. Is this something we could look into? I will be happy to demonstrate it to you in a video meeting or something.